subsys/net/lib/coap/coap.c

```diff
@@ -49,49 +49,117 @@ LOG_MODULE_REGISTER(net_coap, CONFIG_COAP_LOG_LEVEL);
 /* The CoAP message ID that is incremented each time coap_next_id() is called. */
 static uint16_t message_id;
 
-static inline bool append_u8(struct coap_packet *cpkt, uint8_t data)
+static int insert_option(struct coap_packet *cpkt, uint16_t code, const uint8_t *value,
+			 uint16_t len);
+
+static inline void encode_u8(struct coap_packet *cpkt, uint16_t offset, uint8_t data)
+{
+	cpkt->data[offset] = data;
+	++cpkt->offset;
+}
+
+static inline void encode_be16(struct coap_packet *cpkt, uint16_t offset, uint16_t data)
+{
+	cpkt->data[offset] = data >> 8;
+	cpkt->data[offset + 1] = (uint8_t)data;
+	cpkt->offset += 2;
+}
+
+static inline void encode_buffer(struct coap_packet *cpkt, uint16_t offset, const uint8_t *data,
+				 uint16_t len)
+{
+	memcpy(cpkt->data + offset, data, len);
+	cpkt->offset += len;
+}
+
+static bool enough_space(struct coap_packet *cpkt, const uint16_t bytes_to_add)
 {
 	if (!cpkt) {
 		return false;
 	}
 
-	if (cpkt->max_len - cpkt->offset < 1) {
+	if (cpkt->max_len - cpkt->offset < bytes_to_add) {
+		return false;
+	}
+
+	return true;
+}
```

Comment:
This can be an inline oneliner.

```c
static inline bool enough_space(struct coap_packet *cpkt, const uint16_t bytes_to_add)
{
	return (cpkt != NULL) && (cpkt->max_len - cpkt->offset >= bytes_to_add);
}
```

Reply:
imho it's less readable. But it might be more performant